hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,319 @@
+	private final StakeChangeManager manager;
+	private final Supplier<MerkleNetworkContext> networkCtx;
+	private final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo;
+	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts;
+	private final RewardCalculator rewardCalculator;
+	private final SideEffectsTracker sideEffectsTracker;
+	private final GlobalDynamicProperties dynamicProperties;
+
+	private boolean rewardsActivated;
+	private boolean rewardBalanceChanged;
+	private long newRewardBalance;
+
+	private static final Logger log = LogManager.getLogger(StakeAwareAccountsCommitsInterceptor.class);
+	private static final long STAKING_FUNDING_ACCOUNT_NUMBER = 800L;
+
+	public StakeAwareAccountsCommitsInterceptor(
+			final SideEffectsTracker sideEffectsTracker,
+			final Supplier<MerkleNetworkContext> networkCtx,
+			final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo,
+			final GlobalDynamicProperties dynamicProperties,
+			final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts,
+			final RewardCalculator rewardCalculator,
+			final StakeChangeManager manager) {
+		super(sideEffectsTracker);
+		this.manager = manager;
+		this.networkCtx = networkCtx;
+		this.stakingInfo = stakingInfo;
+		this.accounts = accounts;
+		this.rewardCalculator = rewardCalculator;
+		this.sideEffectsTracker = sideEffectsTracker;
+		this.dynamicProperties = dynamicProperties;
+	}
+
+	@Override
+	public void preview(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		// if the rewards are activated previously they will not be activated again
+		rewardsActivated = rewardsActivated || networkCtx.get().areRewardsActivated();
+		rewardBalanceChanged = false;
+
+		super.preview(pendingChanges);
+
+		// The latest period by which an account must have started staking, if it can be eligible for a
+		// reward; if staking is not active, this will return Long.MIN_VALUE so no account is eligible
+		final var latestEligibleStart = rewardCalculator.latestRewardableStakePeriodStart();
+
+		// boolean to track if the account has been rewarded already with one of the pending changes
+		Set<Long> hasBeenRewarded = new HashSet<>();
+
+		// Iterate through the change set, maintaining two invariants:
+		//   1. At the beginning of iteration i, any account that is rewardable due to change in balance or
+		//      stakedAccountId or stakedNodeId or declineRewards fields. Also checks the balance of funding account
+		//      0.0.800 if it has reached the ONE TIME threshold to activate staking.
+		//      NOTE that this activation happens only once.
+		//   2. Any account whose stakedToMe balance is affected by a change in the [0, i) range has
+		//      been, if not already present, added to the pendingChanges; and its changes include its
+		//      new STAKED_TO_ME change
+		for (int i = 0, n = pendingChanges.size(); i < n; i++) {
```

Comment:
Since you encapsulate the second loop below in a helper method `updateNodeStakes()`, it seems symmetric and cleaner to move this loop into a helper method as well.